core/http/server-spring/src/main/java/org/eclipse/rdf4j/http/server/repository/transaction/Transaction.java

```diff
@@ -64,18 +66,54 @@
 
 	private static final Logger logger = LoggerFactory.getLogger(Transaction.class);
 
+	/**
+	 * The period during the {@link #close()} method between waiting for
+	 * {@link ExecutorService#awaitTermination(long, TimeUnit)}, logging that the transaction has not
+	 * completed, and trying again up to {@link #MAX_CLOSE_WAIT_PERIODS} attempts.
+	 */
+	private static final int EXECUTOR_SHUTDOWN_WAIT_PERIOD = 10000;
+
+	/**
+	 * The maximum number of times during the {@link #close()} method that
+	 * {@link ExecutorService#awaitTermination(long, TimeUnit)} will be called before giving up and calling
+	 * {@link ExecutorService#shutdownNow()}.
+	 */
+	private static final int MAX_CLOSE_WAIT_PERIODS = 3;
+
+	/**
+	 * Set to true when entering the {@link #close()} method for the first time, to ensure that only a single
+	 * thread executes the close operations.
+	 */
 	private final AtomicBoolean isClosed = new AtomicBoolean(false);
 
+	/**
+	 * Set to true when the {@link #close()} method is about to complete for the first invocation.
+	 */
+	private final AtomicBoolean closeCompleted = new AtomicBoolean(false);
+
 	private final UUID id;
 
 	private final Repository rep;
 
 	private final RepositoryConnection txnConnection;
 
+	/**
+	 * The {@link ExecutorService} that performs all of the operations related to this Transaction.
+	 */
 	private final ExecutorService executor = Executors.newSingleThreadExecutor(
 			new ThreadFactoryBuilder().setNameFormat("rdf4j-transaction-%d").build());
 
-	private final List<Future<?>> futures = new ArrayList<>();
+	/**
+	 * An append-only {@link Queue} that tracks the state of all of the futures involved in this Transaction.
+	 */
+	private final Queue<Future<?>> futures = new LinkedBlockingQueue<>();
```

Comment:
Sorry, I was going to remove this but forgot. It was done to replace the requirement that results.add be single threaded, but we do actually want add to be atomic with respect to submit, so ArrayList is simpler. LinkedBlockingQueue is the lowest overhead concurrent queue implementation for append-only situations.

Do we want to support removal from results to allow the GC to collect memory if needed, rather than buffering everything in this list? If so, LinkedBlockingDeque, which can efficiently remove from the head of the queue without resizing/moving arrays) may be useful. It could be efficiently done in hasActiveOperations without much overhead. I wasn't previously aware that we had a place in the code that buffered everything (ie, tuple and query results) for a transaction while it was alive, even after the operations were complete.